src/Build/Instance/ProjectInstance.cs

```diff
@@ -641,11 +645,12 @@ private ProjectInstance(ProjectInstance that, bool isImmutable, RequestedProject
                     ProjectItemDefinitionInstance>)this).AfterTargets = CreateCloneDictionary(
                     ((IEvaluatorData<ProjectPropertyInstance, ProjectItemInstance, ProjectMetadataInstance,
                         ProjectItemDefinitionInstance>)that).AfterTargets, StringComparer.OrdinalIgnoreCase);
-                this.TaskRegistry =
-                    that.TaskRegistry; // UNDONE: This isn't immutable, should be cloned or made immutable; it currently has a pointer to project collection
 
-                // These are immutable so we don't need to clone them:
-                this.Toolset = that.Toolset;
+                this.TaskRegistry = that.TaskRegistry.DeepClone();
+                // If the task registry uses toolset identical to the one in project instance - deep clone it just once.
```

Comment:
Same as above please update comment to state that we are reusing already cloned Toolset